test/upgrade-tests.sh

```diff
@@ -0,0 +1,152 @@
+# 1. Still pass, meaning we can upgrade from earlier than latest release (good).
+# 2. Fail, which might be remedied by bumping this version.
+readonly SERVING_RELEASE_YAML=https://github.com/knative/serving/releases/download/v0.2.0/release-no-mon.yaml
+
+function create_serving_latest_release() {
+  echo ">> Bringing up Serving"
+  kubectl apply -f ${SERVING_RELEASE_YAML}
+  # Due to the lack of Status in Istio, we have to ignore failures in initial requests.
+  #
+  # However, since network configurations may reach different ingress pods at slightly
+  # different time, even ignoring failures for initial requests won't ensure subsequent
+  # requests will succeed all the time.  We are disabling ingress pod autoscaling here
+  # to avoid having too much flakes in the tests.  That would allow us to be stricter
+  # when checking non-probe requests to discover other routing issues.
+  #
+  # We should revisit this when Istio API exposes a Status that we can rely on.
+  # TODO(tcnghia): remove this when https://github.com/istio/istio/issues/882 is fixed.
+  echo ">> Patching Istio"
+  kubectl patch hpa -n istio-system knative-ingressgateway --patch '{"spec": {"maxReplicas": 1}}'
+}
+
+function create_everything_latest_release() {
+  # We assume the same istio install as HEAD. Dealing with istio upgrades is
+  # outside of the scope of this test; though this test should catch issues
+  # caused by incompatible istio versions.
+  create_istio
+
+  create_serving_latest_release
+}
+
+function delete_serving_latest_release() {
+  echo ">> Bringing down Serving"
+  kubectl delete --ignore-not-found=true -f ${SERVING_RELEASE_YAML}
+}
+
+# Cribbed from e2e-tests.sh
+function publish_test_images() {
+  echo ">> Publishing test images"
+  image_dirs="$(find ${REPO_ROOT_DIR}/test/test_images -mindepth 1 -maxdepth 1 -type d)"
+  for image_dir in ${image_dirs}; do
+    ko publish -P "github.com/knative/serving/test/test_images/$(basename ${image_dir})"
+  done
+}
+
+function install_latest_release() {
+  # Fail fast during setup.
+  set -o errexit
+  set -o pipefail
+
+  header "Installing latest release"
+  create_everything_latest_release
+
+  # Handle test failures ourselves, so we can dump useful info.
+  set +o errexit
+  set +o pipefail
+
+  wait_until_cluster_up
+}
+
+function install_head() {
+  # Fail fast during setup.
+  set -o errexit
+  set -o pipefail
+
+  header "Installing HEAD"
+  create_everything
+
+  # Handle test failures ourselves, so we can dump useful info.
+  set +o errexit
+  set +o pipefail
+
+  wait_until_cluster_up
+}
+
+# Deletes everything created on the cluster including all knative and istio components.
+function teardown() {
+  delete_everything
+  delete_serving_latest_release
+}
+
+# Script entry point.
+
+initialize $@
+
+header "Setting up environment"
+kubectl create namespace serving-tests
+publish_test_images
+
+install_latest_release
+
+header "Running preupgrade tests"
+report_go_test \
```

Comment:
Once https://github.com/knative/test-infra/pull/234 is in and the vendored copy is updated, this can be simplified to

`go_test_e2e -tags=preupgrade -timeout=20m ./test/upgrade || fail_test`

Same goes for the postupgrade and postdowngrade tests.

Reply:
Trying this I get `go test: tags flag may be set only once` :(

Reply:
Oh sorry I looked at the wrong PR, I see that it's fixed in the PR you linked.